code/modules/food_and_drinks/kitchen_machinery/coffeemaker.dm

```diff
@@ -0,0 +1,693 @@
+/obj/item/storage/box/coffee_cart_rack
+	name = "coffeemaker cartridge box"
+	desc = "A small rack for storing coffeemaker cartridges."
+	var/cartridge_type = /obj/item/coffee_cartridge
+
+/obj/item/storage/box/coffee_cart_rack/Initialize(mapload)
+	. = ..()
+	var/datum/component/storage/STR = GetComponent(/datum/component/storage)
+	STR.max_items = 8
+	STR.can_hold = typecacheof(list(/obj/item/coffee_cartridge))
+
+
+/obj/item/storage/box/coffee_cart_rack/PopulateContents()
+	for(var/i in 1 to 4)
+		new cartridge_type(src)
+		new /obj/item/coffee_cartridge/decaf(src)
+		new /obj/item/coffee_cartridge/fancy(src)
+		new /obj/item/coffee_cartridge(src)
+
+/*
+ * impressa coffee maker
+ * its supposed to be a premium line product, so its cargo-only, the board cant be therefore researched
+ */
+
+/obj/machinery/coffeemaker/impressa
+	name = "impressa coffeemaker"
+	desc = "An industry-grade Impressa Modello 5 Coffeemaker of the Piccionaia Home Appliances premium coffeemakers product line. Makes coffee from fresh dried whole beans."
+	icon = 'icons/obj/machines/coffeemaker.dmi'
+	icon_state = "coffeemaker_impressa"
+	circuit = /obj/item/circuitboard/machine/coffeemaker/impressa
+	initial_cartridge = null //no cartridge, just coffee beans
+	brew_time = 15 SECONDS //industrial grade, its faster than the regular one
+	density = TRUE
+	pass_flags = PASSTABLE
+	/// Current amount of coffee beans stored
+	var/coffee_amount = 0
+	/// List of coffee bean objects are stored
+	var/list/coffee = list()
+
+/obj/machinery/coffeemaker/impressa/Initialize(mapload)
+	. = ..()
+	if(mapload)
+		coffeepot = new /obj/item/reagent_containers/food/drinks/bottle/coffeepot(src)
+		cartridge = null
+
+/obj/machinery/coffeemaker/impressa/Destroy()
+	QDEL_NULL(coffeepot)
+	QDEL_NULL(coffee)
+	return ..()
+
+/obj/machinery/coffeemaker/impressa/examine(mob/user)
+	. = ..()
+	if(coffee)
+		. += span_notice("The internal grinder contains [coffee.len] scoop\s of coffee beans")
+
+/obj/machinery/coffeemaker/impressa/update_overlays()
+	. = ..()
+	. += overlay_checks()
+
+/obj/machinery/coffeemaker/impressa/overlay_checks()
+	. = list()
+	if(coffeepot)
+		if(coffeepot.reagents.total_volume > 0)
+			. += "pot_full"
+		else
+			. += "pot_empty"
+	if(coffee_cups > 0)
+		if(coffee_cups >= max_coffee_cups/3)
+			if(coffee_cups > max_coffee_cups/1.5)
+				. += "cups_3"
+			else
+				. += "cups_2"
+		else
+			. += "cups_1"
+	if(sugar_packs)
+		. += "extras_1"
+	if(creamer_packs)
+		. += "extras_2"
+	if(sweetener_packs)
+		. += "extras_3"
+	if(coffee_amount)
+		if(coffee_amount < 0.7*BEAN_CAPACITY)
+			. += "grinder_half"
+		else
+			. += "grinder_full"
+	return .
+
+/obj/machinery/coffeemaker/impressa/Exited(atom/movable/gone, direction)
+	. = ..()
+	if(gone in coffee)
+		coffee -= gone
+		update_appearance(UPDATE_OVERLAYS)
+
+/obj/machinery/coffeemaker/impressa/try_brew()
+	if(coffee_amount <= 0)
+		balloon_alert_to_viewers("no coffee beans added!")
```

Comment:
Balloon alerts like this shouldn't be shown to viewers
`balloon_alert(user,)` instead

Reply:
done